Document opt_ API

richsalz opened the pull request:
Also made some functions static that didn't need to be public.

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+the OpenSSL command and the internal test programs.
+It is intended to be used like the standard getopt(3) routine, except
+that multi-character flag names are supported, and a variety of parsing
+and other utilities functions are also provided.
```
utility ?

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+set of flags, and the closing C<typedef> name is used for error-checking
+as discussed below.
+
+The B<OPTIONS> C<typedef> specifies an options, what type of argument
```
an = any

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is a the value to return if the option is found. It should be
```
remove 'a'

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+    'M' A signed number that fits in the C<intmax_t> type
+    'U' An unsigned number that fits in the C<uintmax_t> type
+    'l' A signed number that fits in the C<long> type
+    'u' Aunsigned number that fits in the C<unsigned long> type
```
A unsigned

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+    'n' A signed number that fits in the C<int> type
+    'p' A non-negative number that fits in the C<int> type
+    'M' A signed number that fits in the C<intmax_t> type
+    'U' An unsigned number that fits in the C<uintmax_t> type
```
Choose 'An' or 'A' for these

mattcaswell replied:
Errrr....no? What @richsalz has here is correct. "An unsigned" but "A" for everything else.

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+should not appear in the array.
+
+If the help string is too long to fit into one line, it may be continued
+on multiple lines each entry should use B<OPT_MORE_STR>, like this:
```
, each

slontis commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+         "This flag is not really needed on Unix systems"}
+        {OPT_MORE_STR, 0, 0,
+
+Each subsequent line will be indent the correct amount.
```
indented?

levitte commented on doc/internal/man3/opt.pod:
```diff
@@ -0,0 +1,182 @@
+
+=head1 SYNOPSIS
+
+ #include "apps/include/opt.h"
```
We never include it like that.  Rather, we compile with `-I PATH/TO/apps/include` and then have `#include "opt.h"`

richsalz commented:
Thanks @slontis and @levitte .  Fixup commit pushed addressing your feedback. @mattcaswell was right about me being right on the global A/An change :)  I just marked things as resolved to reduce the email.  Now to finish it.

richsalz commented:
Ready for review now. @ddvo, sorry about the "not apps/cms.c" comment, but it *is* true.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+        OPT_YES, OPT_NAME, OPT_COUNT, OPT_OFILE
+    } OPTION_CHOICE;
+
+The first two lines must appear exactly as shown. The third line is a sample
```
Good that this is documented here.
Yet recently I have been wondering why each app is expected to define 
```
        OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
```
which adds just redundancy and the risk of inconsistency.

richsalz replied:
The `opt_next` function returns -1 on error and 0 when done.  Therefore both those values should be in the enumeration of possible values that the application is expecting.  You can cast the `opt_next` value to your enumeration type, and the compiler can warn you if you're missing cases.  That's done in the `OPT_x_CASES` code handling.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is the value to return if the option is found. It should be
+one of the choices in the enumeration above.
```
For the retvals it should also be mentioned that they are expected to be unique (though I'm going to add an exception to this rule soon) and the uniqueness is checked at runtime.

richsalz replied:
No, the retval does not have to be unique.  It's how we alias options.  The *name* of the option has to be unique.
For example, the "enc" command  has
```
    {"a", OPT_A, '-', "Base64 encode/decode, depending on encryption flag"},
    {"base64", OPT_A, '-', "Same as option -a"},
```

I assume you're doing something like that.

DDvO replied:
Ah, I confused the two, so my comment was actually about the option names.
These are checked for uniqueness.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+    '<' Name of file to open for input
+    '>' Name of file to open for output
+    'n' A signed number that fits in the C<int> type
+    'p' A nonnegative number that fits in the C<int> type
```
nonnegative -> positive

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+    'E' Like 'F' but also allows ENGINE
+    'f' Any file format
+
+The B<helpstr> is what to display when the user uses the help flag,
```
"help flag" or "help option"?

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+The B<helpstr> is what to display when the user uses the help flag,
+which should be C<"help">.
+
+A program should declare its options right after the enumeration:
```
Please mention that the order of the entries should be the same as in the enumeration, which helps readability and maintainability.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+the value from the constants in F<fmt.h> file.
+
+The opt_string() function checks that B<name> appears in the
+B<NULL>-terminated array of strings. It returns B<1>ne if found,
```
Funny style: `B<1>ne`

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,269 @@
+
+The best examples can be found in sources for the commands in the F<apps>
+directory of the source tree.
+A notable exception is F<apps/cmp.c> which uses this API, but does
```
No problem mentioning this.
This CMP app allows putting all options also in a config file in addition to the command line, while the options on the command line can override those given in the config file.

DDvO reviewed (commented):
Nice work.
I just added some quick comments.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -246,16 +254,41 @@ left, and opt_rest() returns a pointer to the first non-option.
+
+The B<OPT_x_ENUM> macro is used to define the numeration values, where B<x>
+is one of the letters above.  The B<OPT_x_OPTIONS> macro is used to
+list the set of common options, and the B<OPT_R_CASES> is used in
```
OPT_x_CASES

richsalz replied:
fixup commit pushed, thanks.

beldmit commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,302 @@
+that is the set of options the program accepts. For example:
+
+    typedef enum OPTION_choice {
+        OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
```
If this set of options must appear exactly this, doesn't it mean that it's worth converting them to smth like verification options?

DDvO replied:
A very simple solution would be to turn
```
         OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
```
into a macro.

richsalz replied:
Turning those common options into a macro is for another PR,.  This is purely documentation.  FWIW, I disagree with that idea and won't be doing it :)

DDvO replied:
That's why I wrote: would ;-)

richsalz commented:
fixup: non-negative -> nonnegative

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+    const char *helpstr;
+
+The B<name> is the name of the option that the user would type. Options
+are words preceeded with a minus sign. If the user uses two minus signs,
```
preceeded -> preceded

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+    'U' An unsigned number that fits in the C<uintmax_t> type
+    'l' A signed number that fits in the C<long> type
+    'u' An unsigned number that fits in the C<unsigned long> type
+    'c' File of fomer PEM, DER, or S/MIME
```
former

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+There are a several useful functions for parsing numbers.  These are
+opt_int(), opt_long(), opt_ulong(), opt_imax(), and opt_umax().  They all
+take C<0x> to mean hexadecimal and C<0> to mean octal, and will do the
+necessary range-checking. They return B<1> if succcessful and fill in the
```
successful

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+opt_num_rest, opt_rest, opt_help, opt_isdir
+- Option parsing for commands and tests
+
+=head1 SYNOPSIS
+
```
The following symbols are not (yet) documented:
```
const char *opt_path_end(const char *filename);
void opt_begin(void);
const char *format2str(int format);
int opt_int_arg(void);
char *opt_name(void);
int opt_verify(int i, X509_VERIFY_PARAM *vpm);
int opt_rand(int i);
int opt_provider(int i);
void opt_print(const OPTIONS * opt, int doingparams, int width);
int opt_format_error(const char *s, unsigned long flags);
void print_format_error(int format, unsigned long flags);
int opt_printf_stderr(const char *fmt, ...);
```

richsalz replied:
For another PR for someone else to do.  What I wrote is good enough for a new openssl developer to write basic options.  (And `opt_int_arg` should crawl away to somewhere else :)

DDvO replied:
Fine by me - the given coverage of functions should already be very helpful.

DDvO reviewed (commented):
A couple of typos and a bunch of missing functions

DDvO commented:
Ah, and there is the wonderful type `STRINT_PAIR` - I suggest removing it on this occasion
(replacing all uses by its synonym `OPT_PAIR` - I tried it out, it works despite the comment in `opt.h`) - 
then there is no reason to document it :wink:

richsalz commented:
Re: STRINT_PAIR:  speak to @paulidale about that one.

I maintain this is good enough to accept for internal documentation right now.  I'm not adding new content.

richsalz commented:
fixup commit pushed to fix typo's.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+
+Within the option parsing loop, opt_flag() returns the option,
+without any leading hyphens. The opt_arg() function returns
+the option's value, if there is one..
```
double period at EOL

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+or prints a diagnostic and returns B<0> if not.
+
+The opt_pair() function takes a list of B<pairs>, each of which
+has a text name and numeric value. The specified B<name> is
```
numeric -> an integer

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+
+The common options are used throughout the sources for the OpenSSL commands.
+They are also used with common descriptions when generating the
+manpages, in the file C<doc/perlvars.pm>, which follow a similar naming
```
C<doc/perlvars.pm> -> F<doc/perlvars.pm>

DDvO reviewed (changes requested):
Just a few more nits, then I'll approve.

richsalz commented:
updated commit pushed fixing all just-noticed typo's :)

DDvO reviewed (approved):
LGTM

t8m commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+the value from the constants in F<fmt.h> file.
+
+The opt_string() function checks that B<name> appears in the
+B<NULL>-terminated array of strings. It returns B<1> if found,
```
Remove `B<>` around NULL. Also I am not sure about bolding the constants such as 0 and 1.

richsalz replied:
Well it's not obvious :)
```
; g grep NULL doc | grep 'B<'|wc
    461    4893   43517
; g grep NULL doc | grep -v 'B<'|wc
   1362   11938  117044
```

But I removed the `B<>` around NULL and the 0/1 return values.

t8m commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+B<NULL>-terminated array of strings. It returns B<1> if found,
+or prints a diagnostic and returns B<0> if not.
+
+The opt_pair() function takes a list of B<pairs>, each of which
```
Parameters such as `pairs` here should be in `I<>` and not `B<>`.

t8m reviewed (changes requested):
Please fix the markup.

richsalz commented:
fixup commit pushed to address the markup nits that @t8m found.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+    {OPT_HELP_STR, 1, '-', Usage: %s [options] [text...]\n"}
+
+The B<retval> and B<valtype> are ignored, and the B<helpstr> should
+follow the general construction as shown. The B<%s> will get the program
```
C<%s>

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+In an option list, there can be at most one option with the empty string.
+This is a "wildcard" or "unknown" option. For example, it allows an
+option to be be taken as digest algorithm, like B<-sha1>. The
+function opt_cipher() takes the specified B<name> and fills in
+the cipher into B<cipherp>.  The function opt_md() does the same
+thing for message digest.
```
For all funciton arguments, throughout the doc, use I<> rather than B<>.

DDvO reviewed (changes requested):
Some more markup nits.

richsalz commented:
YAFCP  (yet another fixup commit pushed)

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -223,9 +223,9 @@ the option's value, if there is one.
-option to be be taken as digest algorithm, like B<-sha1>. The
-function opt_cipher() takes the specified B<name> and fills in
-the cipher into B<cipherp>.  The function opt_md() does the same
+option to be be taken as digest algorithm, like C<-sha1>. The
```
YAN (yet another nit): 
B<-sha> was correct.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+    int valtype;
+    const char *helpstr;
+
+The B<name> is the name of the option that the user would type. Options
```
I<name>

DDvO replied:
There are still many more such func arg names below.

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is the value to return if the option is found. It should be
```
I<

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+After processing all the options, the opt_num_rest() returns what is
+left, and opt_rest() returns a pointer to the first non-option.
+If there were no parameters, it will point to the NULL that is
+at the end of the standard B<argv> array.
```
I<

DDvO commented on doc/internal/man3/OPTIONS.pod:
```diff
@@ -0,0 +1,301 @@
+The B<retval> is the value to return if the option is found. It should be
+one of the choices in the enumeration above.
+
+The B<valtype> defines what the option's parameter must be. It should
```
I<

richsalz commented:
I'm done.  Feel free to push your own fixup commits to this branch.

DDvO reviewed (changes requested):
The last fixup was incomplete.

DDvO commented:
> YAFCP (yet another fixup commit pushed)

This would make a nice CLI command alias :rofl:

richsalz commented:
> The last fixup was incomplete.

I encourage you to make the changes.  As I said, for this *INTERNAL* manpage I'm done.

DDvO commented:
> > The last fixup was incomplete.
> 
> I encourage you to make the changes. As I said, for this _INTERNAL_ manpage I'm done.

Come on, just make the markup of function arg names consistent 
and I'll re-approve. And I presume @t8m will be happy as well.

DDvO commented:
Maybe we should have a doc-nit-pick tool analogous to `check-format.pl` :wink:

richsalz commented:
I made all the function arg names use `I<>`  You're now also asking for field names to be changed, if I understand comments like the one at line 96. And I shouldn't have to spend time trying to understand.  Shrug.  Change it, merge it, reject it. The nits have worn me down.

DDvO commented:
> I made all the function arg names use `I<>` You're now also asking for field names to be changed, if I understand comments like the one at line 96. And I shouldn't have to spend time trying to understand. Shrug. 

Oops, right, this is an instance of a field name! For them I'm not sure what the convention is - @t8m?
I just know that, these days, func arg names are supposed to be written like this: `I<arg>`
since I was often asked to make changes in this direction.

`B<argv>` is a function arg name, so it should look like this: `I<argv>`.
Similarly, `"argc, argv"` would better be given as `I<argc>, I<argv>`.
And the CLI option `I<-sha1>` should be rendered as `B<-sha1>`.
Likely these four nits are the only ones remaining.

t8m commented:
I'm willing to approve anyway as this is internal API doc.

DDvO reviewed (approved):
So I'm approving, too. 
Good enough for now; we can fix the remaining nits later.
Thanks @richsalz for taking up documenting `opt.h`.

richsalz commented:
Thanks for the comments and feedback, glad this is getting merged.  Hope someone will fix all the remaining nits at some point, I guess.  (And yes, as the original author of find-doc-nits I appreciate the irony)

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Squashed and merged :)

DDvO commented:
The issues identified above and not solved by this PR are now handled in the follow-up PR #15111.
